relay-cardinality/src/redis.rs

```
@@ -0,0 +1,427 @@
+    {
+        let entries = entries.into_iter().into_group_map_by(|f| Scope {
+            organization_id,
+            namespace: f.namespace,
+        });
+
+        let timestamp = SystemTime::now()
+            .duration_since(SystemTime::UNIX_EPOCH)
+            .unwrap_or_default()
+            .as_secs();
+
+        // Allows to fast forward time in tests.
+        #[cfg(test)]
+        let timestamp = timestamp + self.time_offset;
+
+        let mut client = self.redis.client()?;
+        let mut client = RedisHelper(client.connection()?);
+
+        let mut accepted = Vec::new();
+        for (scope, entries) in entries {
+            accepted.push(self.check_limits(&mut client, scope, entries, timestamp, limit)?);
+        }
+
+        Ok(Box::new(accepted.into_iter().flatten()))
+    }
+}
+
+/// Internal Redis client wrapper to abstract the issuing of commands.
+struct RedisHelper<'a>(Connection<'a>);
+
+impl<'a> RedisHelper<'a> {
+    fn read_set(&mut self, name: &str) -> Result<HashSet<u32>> {
+        let result = redis::cmd("SMEMBERS")
+            .arg(name)
+            .query(&mut self.0)
+            .map_err(relay_redis::RedisError::Redis)?;
+
+        Ok(result)
+    }
+
+    fn add_to_set<I: ToRedisArgs + std::fmt::Debug>(
+        &mut self,
+        name: &str,
+        values: impl IntoIterator<Item = I>,
+        expire: u64,
+    ) -> Result<()> {
+        for values in &values.into_iter().chunks(REDIS_MAX_SADD_ARGS) {
+            let mut cmd = redis::cmd("SADD");
+            cmd.arg(name);
+
+            for arg in values {
+                cmd.arg(arg);
+            }
+
+            cmd.query(&mut self.0)
+                .map_err(relay_redis::RedisError::Redis)?;
+        }
+
+        redis::cmd("EXPIRE")
+            .arg(name)
+            .arg(expire)
+            // .arg("NX") - NX not supported for Redis < 7
+            .query(&mut self.0)
+            .map_err(relay_redis::RedisError::Redis)?;
```

Will this raise an error if the set was removed between redis commands? Or rather fail silently?

---

I double checked, this fails silently and only raises an error when there is an issue with redis/the connection.